actors/verifreg/src/emit.rs

```diff
@@ -0,0 +1,86 @@
+// A namespace for helpers that build and emit verified registry events.
+
+use crate::{ActorError, Allocation, AllocationID, Claim};
+use crate::{ClaimID, DataCap};
+use fil_actors_runtime::runtime::Runtime;
+use fil_actors_runtime::EventBuilder;
+use fvm_shared::ActorID;
+
+/// Indicates a new value for a verifier's datacap balance.
+/// Note that receiving this event does not necessarily mean the balance has changed.
+/// The value is in datacap whole units (not TokenAmount).
+pub fn verifier_balance(
+    rt: &impl Runtime,
+    verifier: ActorID,
+    new_balance: &DataCap,
+) -> Result<(), ActorError> {
+    rt.emit_event(
+        &EventBuilder::new()
+            .event_type("verifier-balance")
+            .field_indexed("verifier", &verifier)
+            .field("balance", new_balance)
+            .build()?,
+    )
+}
+
+/// Indicates a new allocation has been made.
+pub fn allocation(
+    rt: &impl Runtime,
+    id: AllocationID,
+    alloc: &Allocation,
+) -> Result<(), ActorError> {
+    rt.emit_event(&EventBuilder::new().event_type("allocation").with_allocation(id, alloc).build()?)
+}
+
+/// Indicates an expired allocation has been removed.
+pub fn allocation_removed(
+    rt: &impl Runtime,
+    id: AllocationID,
+    alloc: &Allocation,
+) -> Result<(), ActorError> {
+    rt.emit_event(
+        &EventBuilder::new().event_type("allocation-removed").with_allocation(id, alloc).build()?,
+    )
+}
+
+/// Indicates an allocation has been claimed.
+pub fn claim(rt: &impl Runtime, id: ClaimID, claim: &Claim) -> Result<(), ActorError> {
+    rt.emit_event(&EventBuilder::new().event_type("claim").with_claim(id, claim).build()?)
+}
+
+/// Indicates an existing claim has been updated (e.g. with a longer term).
+pub fn claim_updated(rt: &impl Runtime, id: ClaimID, claim: &Claim) -> Result<(), ActorError> {
+    rt.emit_event(&EventBuilder::new().event_type("claim-updated").with_claim(id, claim).build()?)
+}
+
+/// Indicates an expired claim has been removed.
+pub fn claim_removed(rt: &impl Runtime, id: ClaimID, claim: &Claim) -> Result<(), ActorError> {
+    rt.emit_event(&EventBuilder::new().event_type("claim-removed").with_claim(id, claim).build()?)
+}
+
+// Private helpers //
+trait WithAllocation {
+    fn with_allocation(self, id: AllocationID, alloc: &Allocation) -> EventBuilder;
+}
+
+impl WithAllocation for EventBuilder {
+    fn with_allocation(self, id: AllocationID, alloc: &Allocation) -> EventBuilder {
+        self.field_indexed("id", &id)
+            .field_indexed("client", &alloc.client)
+            .field_indexed("provider", &alloc.provider)
+            .field_indexed("data-cid", &alloc.data)
```

Comment:
Based on https://github.com/filecoin-project/FIPs/discussions/754#discussioncomment-7204690, I don't think we should add any fields beyond ID. Which also means you can remove this whole method. Ditto for claims.

Reply:
@anorth Is an Allocation Id unique across ALL allocations ? Same question for claims.

Reply:
@anorth Didn't realise that an Allocation/Claim ID is unique across all allocation/claims. If that is the case, why do we need more fields for looking up a claim/allocation here in `get_allocation` AND `get_claim` here ?

https://github.com/filecoin-project/builtin-actors/blob/1e50065123c05fd4881cc1898d59666c3bb97233/actors/verifreg/src/state.rs#L240

Reply:
Yes they are unique, but they are not indexed directly. The only lookup available is via (client, allocation) or (provider, claim).